Library/Homebrew/readall.rb

```diff
@@ -39,19 +43,27 @@ def valid_aliases?(alias_dir, formula_dir)
       !failed
     end
 
-    def valid_formulae?(formulae, bottle_tag: nil)
-      success = T.let(true, T::Boolean)
-      formulae.each do |file|
-        base = Formulary.factory(file)
-        next if bottle_tag.blank? || !base.path.exist? || !base.class.on_system_blocks_exist?
-
-        formula_contents = base.path.read
+    def valid_formulae?(tap, bottle_tag: nil)
+      cache[:valid_formulae] ||= {}
 
-        readall_namespace = Formulary.class_s("Readall#{bottle_tag.to_sym.capitalize}")
-        readall_formula_class = Formulary.load_formula(base.name, base.path, formula_contents, readall_namespace,
-                                                       flags: base.class.build_flags, ignore_errors: true)
-        readall_formula_class.new(base.name, base.path, :stable,
-                                  alias_path: base.alias_path, force_bottle: base.force_bottle)
+      success = T.let(true, T::Boolean)
+      tap.formula_files.each do |file|
+        valid = cache[:valid_formulae][file]
+        next if valid == true || valid&.include?(bottle_tag)
+
+        formula_name = file.basename(".rb").to_s
+        formula_contents = file.read(encoding: "UTF-8")
+
+        readall_namespace = "ReadallNamespace"
+        readall_formula_class = Formulary.load_formula(formula_name, file, formula_contents, readall_namespace,
+                                                       flags: [], ignore_errors: false)
+        readall_formula = readall_formula_class.new(formula_name, file, :stable, tap: tap)
+        readall_formula.to_hash
+        cache[:valid_formulae][file] = if readall_formula.on_system_blocks_exist?
+          [bottle_tag, *cache[:valid_formulae][file]]
+        else
+          true
+        end
```

Comment:
```suggestion
        unless readall_formula.on_system_blocks_exist?
          cache[:valid_formulae][file] = true
        end
```

I don't think we really need to track the case where a formula uses on system blocks. We only run this method once per os/arch combination anyway.

The main thing we're trying to do here is avoid reading formula without on system blocks multiple times (since they shouldn't change), right?

```
[2] brew(main)> CoreTap.instance.formula_files.count do |file|
[2] brew(main)*   Formulary.factory(file).on_system_blocks_exist?
[2] brew(main)* end
=> 880
[3] brew(main)> CoreTap.instance.formula_files.count
=> 6781
```

Isn't it possible to use the manual os and arch functions in formula definitions though like `OS.mac?`? Or maybe that's only limited to certain stanzas like `install` and `test` that aren't relevant for `brew readall`. To be fair this command was update recently to test against different os/arch combinations but it does change the behavior, right?

Reply:
We don't actually allow `on_system` blocks inside `install` and `test`. We do however in `caveats`, which is read by `to_hash`.

Reply:
Sounds good. Do you think this sort of optimization would be useful for the cask repo as well at some point?